src/V3Delayed.cpp

```diff
@@ -76,140 +114,174 @@
 VL_DEFINE_DEBUG_FUNCTIONS;
 
 //######################################################################
-// Delayed state, as a visitor of each AstNode
+// Convert AstAssignDlys (NBAs)
 
 class DelayedVisitor final : public VNVisitor {
     // TYPES
 
-    // Components of an AstAssignDly LHS expression, e.g.:
-    //   Sel(ArraySel(ArraySel(_: VarRef, _: Index), _: Index), _: Lsb, _: Width))
-    struct LhsComponents final {
-        AstVarRef* refp = nullptr;  // The referenced variable
-        std::vector<AstNodeExpr*> arrIdxps;  // The array indices applied to 'refp'
-        AstNodeExpr* selLsbp = nullptr;  // The bit select LSB expression, after the array selects
-        AstConst* selWidthp = nullptr;  // The width of the bit select
+    // The various NBA conversion schemes, including error cases
+    enum class Scheme : uint8_t {
+        Undecided = 0,
```

Comment:
Should these be classes instead with virtual functions, then pass around a pointer to that scheme instead of enum value?  Would simplify some of your case-statement dispatching code, also then Partial/Whole could be specializations of a common base class.  But I don't know if that's all cleaner in the end.

Reply:
That's kind of what I thought initially, but those classes would depend on shared state/functions in the `DelayedVisitor` class, so would not be clearer with all the cross referencing necessary to do it that way. Also this contains markers for the unsupported cases.